### ISSUE TITLE
uploadProduct fails

### ISSUE THREAD

**Issue by polluks** [otto-de/marketplace-php-sdk#2]
**Describe the bug**
I download a product, it works as expected.
I upload the same product and it fails.

**To Reproduce**
`php.exe SampleCli.php -e live -u api_*** -p *** --loglevel debug uploadProduct 5058-orig.json`

**Expected behavior**
No state failed of course.
No one minute pending every time.

**Screenshots**
````
No HTTP client provided for Oauth2ApiAccessor, constructing default Guzzle client
Setting up the Oauth2ApiAccessor
Generating initial access token
Received access token valid until 1625216828
Parsed 1 product variations
Creating new client for partner products
Posting next batch of 1product variations
Making authenticated POST request to https://live.api.otto.market/v1/products
Access token is valid (until 1625216828)
POST returned with status code 202
Uploaded product variations.
````
[...] after one minute
````
Did not update upload progress - pingAfter not yet reached
{
    "state": "pending",
    "message": "The process is currently in progress",
    "progress": 0,
    "total": 1,
    "pingAfter": "2021-07-02T08:38:09+00:00",
    "links": [
        {
            "rel": "self",
            "href": "\/v1\/products\/update-tasks\/9498a13b-f6c9-4445-b1c5-3f3b56fb5a4f"
        },
        {
            "rel": "failed",
            "href": "\/v1\/products\/update-tasks\/9498a13b-f6c9-4445-b1c5-3f3b56fb5a4f\/failed"
        },
        {
            "rel": "succeeded",
            "href": "\/v1\/products\/update-tasks\/9498a13b-f6c9-4445-b1c5-3f3b56fb5a4f\/succeeded"
        }
    ]
}
Making authenticated GET request to https://live.api.otto.market/v1/products/update-tasks/9498a13b-f6c9-4445-b1c5-3f3b56fb5a4f
Access token is valid (until 1625216828)
Updated upload progress
Making authenticated GET request to https://live.api.otto.market/v1/products/update-tasks/9498a13b-f6c9-4445-b1c5-3f3b56fb5a4f/failed
Access token is valid (until 1625216828)
{
  "results": [
    {
      "variation": "/v1/products/5058AR-001",
      "errors": [
        {
          "code": "300003 - INVALID_MEDIA_ASSET_LINK",
          "title": "Der Link ist ungültig",
          "path": "/products",
          "jsonPath": "$.productVariations[?(@.sku==\"5058AR-001\")].mediaAssets"
        }
      ]
    }
  ]
}
````

**Desktop:**
 - OS: Windows 10
 - PHP 8.0.7 (cli) (built: Jun  2 2021 00:41:03) ( ZTS Visual C++ 2019 x64 )

**Comment by marcogreiveldinger** [comment]
Hi @polluks,

thanks for describing your issue. Am I understanding correctly that you are using the sample command line interface to upload a product?
As the name already says this is an example on how you could use the sdk itself. You can easily customize the behaviour of this sample.

You could react on the state of the uploadStatus like that for example:
```php
  case 'uploadProduct':
    $fileContents = file_get_contents($options->getArgs()[0]);
    $productVariations = ObjectSerializer::deserialize($fileContents, '\Otto\Market\Products\Model\ProductVariation[]');
    $this->debug('Parsed ' . sizeof($productVariations) . ' product variations');
    $lastStatus = $this->productClient()->postProducts($productVariations);
    $this->info('Uploaded product variations.');
    foreach ($lastStatus as $status) {
        $this->info($status);
        while ($status->getState() == ProductProcessProgress::STATE_PENDING) {
            sleep(30);
            $status = $this->productClient()->getUpdatedUploadProgress($status);
            $this->info('State: ' . $status->getState());
        }
    }
    break;
```

The mentioned one minute of waiting between the requests is intended as the product upload replies with a proposed timestamp to ask again for a new status update.

**Comment by marcogreiveldinger** [comment]
After a bit of research I discovered the bug why the upload is failing:

When a product is managed via the Otto Partner Connect Portal (OPC) then the images are uploaded via a direct file upload in the browser. This results in the product having a mediaAsset container with no location(url to the image), as this image is not downloaded asynchronously.

When now getting this exact product via api the original filename is returned instead of a location. As the SDK is based on the public openapi documentation (and there is no filename mentioned in there yet) the sdk is not able to deserialize it as it should. After for example modifying the product description, the upload fails at the api because the mediaAsset container is not valid.

For now you could try a complete new upload which should work!

**Comment by 5l1v3r1** [comment]
php SampleCli.php -e live -u 'api_*****' -p '******' --loglevel debug uploadProduct products.json
No HTTP client provided for Oauth2ApiAccessor, constructing default Guzzle client
Setting up the Oauth2ApiAccessor
Generating initial access token
Received access token valid until 1626807687
InvalidArgumentException caught in /mnt/c/Users/User/Desktop/SON/vendor/otto/marketplace-php-sdk/generated/products/ObjectSerializer.php:279
#0 /mnt/c/Users/User/Desktop/SON/vendor/SampleCli.php(104): Otto\Market\Products\ObjectSerializer::deserialize()
#1 /mnt/c/Users/User/Desktop/SON/vendor/splitbrain/php-cli/src/CLI.php(175): Otto\Market\Cli\Client\SampleCli->main()
#2 /mnt/c/Users/User/Desktop/SON/vendor/splitbrain/php-cli/src/CLI.php(97): splitbrain\phpcli\CLI->execute()
#3 /mnt/c/Users/User/Desktop/SON/vendor/SampleCli.php(200): splitbrain\phpcli\CLI->run()
#4 {main}
☠ Invalid array '\Otto\Market\Products\Model\ProductVariation[]'

I have similar problem. I used the products json from the shopify provied list. (api)

**Comment by 5l1v3r1** [comment]
Ok i commented out the related part at Objectserializer and it seems to work but there is another problem now. What kind of json list does this cli accept? Obviously it does not work with products.json of Shopify. If you can explain properly i can map the properties from shopify's products.json to another acceptable json list which is compatible with your script.

**Comment by marcogreiveldinger** [comment]
Hi @5l1v3r1,

could you provide me an example of yours(especially the shopify api result)? Then I could have a closer look why exactly it is failing.
As I already stated above the CLI is only an example on how the sdk itself could be used. You could easily write your own CLI which fits your needs more.

The CLI should accept a list of ProductVariations. An example can be found at the [otto market api page ](https://api.otto.market/docs/03_Products/v2/products-interface.html#/Products/createOrUpdateProductVariations)


<details>
  <summary>Unfold for an example</summary>

```
[
  {
    "productReference": "UBN-11779",
    "sku": "3858389911564",
    "ean": "3858389911564",
    "isbn": "978-3-16-148410-0",
    "upc": "042100005264",
    "pzn": "PZN-4908802",
    "mpn": "H2G2-42",
    "moin": "93992000200",
    "offeringStartDate": "2019-10-19T09:30:00.000Z",
    "releaseDate": "2019-10-19T09:30:00.000Z",
    "maxOrderQuantity": 5,
    "productDescription": {
      "category": "Outdoorjacke",
      "brand": "Adidas",
      "productLine": "501",
      "manufacturer": "3M",
      "productionDate": "2021-07-02T09:30:52.093Z",
      "multiPack": true,
      "bundle": false,
      "fscCertified": true,
      "disposal": false,
      "productUrl": "http://myproduct.somewhere.com/productname/",
      "description": "<p>Some example words...<b>in bold</b>...some more</p>",
      "bulletPoints": [
        "My top key information..."
      ],
      "attributes": [
        {
          "name": "Bundweite",
          "values": [
            "34"
          ],
          "additional": true
        }
      ]
    },
    "mediaAssets": [
      {
        "type": "IMAGE",
        "location": "http://apartners.url/image-location"
      }
    ],
    "delivery": {
      "type": "PARCEL",
      "deliveryTime": 1
    },
    "pricing": {
      "standardPrice": {
        "amount": 19.95,
        "currency": "EUR"
      },
      "vat": "FULL",
      "msrp": {
        "amount": 19.95,
        "currency": "EUR"
      },
      "sale": {
        "salePrice": {
          "amount": 19.95,
          "currency": "EUR"
        },
        "startDate": "2019-10-19T09:30:00.000Z",
        "endDate": "2019-10-19T09:30:00.000Z"
      },
      "normPriceInfo": {
        "normAmount": 100,
        "normUnit": "g",
        "salesAmount": 500,
        "salesUnit": "g"
      }
    },
    "logistics": {
      "packingUnitCount": 3,
      "packingUnits": [
        {
          "weight": 365,
          "width": 600,
          "height": 200,
          "length": 300
        }
      ]
    }
  },
  {
    "productReference": "UBN-11779",
    "sku": "3858389911564",
    "ean": "3858389911564",
    "isbn": "978-3-16-148410-0",
    "upc": "042100005264",
    "pzn": "PZN-4908802",
    "mpn": "H2G2-42",
    "moin": "93992000200",
    "offeringStartDate": "2019-10-19T09:30:00.000Z",
    "releaseDate": "2019-10-19T09:30:00.000Z",
    "maxOrderQuantity": 5,
    "productDescription": {
      "category": "Outdoorjacke",
      "brand": "Adidas",
      "productLine": "501",
      "manufacturer": "3M",
      "productionDate": "2021-07-02T09:30:52.093Z",
      "multiPack": true,
      "bundle": false,
      "fscCertified": true,
      "disposal": false,
      "productUrl": "http://myproduct.somewhere.com/productname/",
      "description": "<p>Some example words...<b>in bold</b>...some more</p>",
      "bulletPoints": [
        "My top key information..."
      ],
      "attributes": [
        {
          "name": "Bundweite",
          "values": [
            "34"
          ],
          "additional": true
        }
      ]
    },
    "mediaAssets": [
      {
        "type": "IMAGE",
        "location": "http://apartners.url/image-location"
      }
    ],
    "delivery": {
      "type": "PARCEL",
      "deliveryTime": 1
    },
    "pricing": {
      "standardPrice": {
        "amount": 19.95,
        "currency": "EUR"
      },
      "vat": "FULL",
      "msrp": {
        "amount": 19.95,
        "currency": "EUR"
      },
      "sale": {
        "salePrice": {
          "amount": 19.95,
          "currency": "EUR"
        },
        "startDate": "2019-10-19T09:30:00.000Z",
        "endDate": "2019-10-19T09:30:00.000Z"
      },
      "normPriceInfo": {
        "normAmount": 100,
        "normUnit": "g",
        "salesAmount": 500,
        "salesUnit": "g"
      }
    },
    "logistics": {
      "packingUnitCount": 3,
      "packingUnits": [
        {
          "weight": 365,
          "width": 600,
          "height": 200,
          "length": 300
        }
      ]
    }
  }
]
```

</details>

**Comment by 5l1v3r1** [comment]
Ok. Thank you for response. Here i add both products.json of shopify and also i mapped the attributes via omega app at our shop to create otto-de.csv and then i converted this csv to json. Both of them seem like not working.

[products.zip](https://github.com/otto-de/marketplace-php-sdk/files/6977984/products.zip)

**Comment by marcogreiveldinger** [comment]
@5l1v3r1 the mapped otto-de.json is not far away from working but here is the error!

Compare the example I posted above against your otto-de.json, then you should clearly see what have gone wrong.

For example make sure that attributes like _mediaAssets, logistics, pricing, productDescription, delivery_ are containers which are holding a subobject.
And you can leave attributes which are empty out, so you do not have a cluttered json.

Have a look again at the official api documentation [link](https://api.otto.market/docs/03_Products/v2/products-interface.html#/Products/createOrUpdateProductVariations) where all objects are described and examples are provided

**Comment by 5l1v3r1** [comment]
Yes i understand what you meant but it seems impossible to get the exact mapping with our omega mapping app and unfortunately there is no other mapping app that provides us mapping from shopify's products.json to customized json files. In fact they lack this option for a long time. What do you suggest me to do? I dont understand why otto does not allow bulk upload via csv files through their web interface since we have thousands of product variants and we cannot add them manually one by one. To be honest we checked other integration platforms for otto.de and the price is very very expensinve just for one integration. Can you help us with that? At least you can show us how to convert our products.json file to otto.de acceptable json files.

**Comment by torstenbeuck0815** [comment]
This issue is fixed in Release 0.2.3
